File: 354-russian-doll-envelopes/russian-doll-envelopes.java
class Solution {

    private int LIS(int[] heights) 
    {
        int n = heights.length;
        TreeSet<Integer> treeSet = new TreeSet<>();
        for(int i = 0; i < n; i++)
        {
            int num = heights[i];
            Integer ceil = treeSet.ceiling(num);

            if(ceil != null)
            treeSet.remove(ceil);

            treeSet.add(num);
        }
        return treeSet.size();
    }

    public int maxEnvelopes(int[][] envelopes) 
    {
        Arrays.sort(envelopes, (a, b) -> {
            return a[0] == b[0] ? b[1] - a[1] : a[0] - b[0];
        });

        int index = 0;
        int[] heights = new int[envelopes.length];
        for(int[] envelope : envelopes)
        heights[index++] = envelope[1];

        return LIS(heights);
    }
}

File: 354-russian-doll-envelopes/russian-doll-envelopes.cpp
class Solution {
public:
    int LIS(vector<vector<int>>& env)
    {
        int n = env.size();
        set<long long> treeSet;
        // Apply LIS on heights
        for(int i = 0; i < n; i++)
        {
            int width = env[i][1];
            if(i > 0 && width == env[i - 1][1])
            continue;

            auto it = treeSet.lower_bound(width);

            if(it != treeSet.end())
            treeSet.erase(*it);

            treeSet.insert(width);
        }
        return treeSet.size();
    }

    int maxEnvelopes(vector<vector<int>>& envelopes) 
    {
        // Step 1: Sort by width ASC, height DESC
        sort(envelopes.begin(), envelopes.end(), [](const vector<int>& a, const vector<int>& b){
            return a[0] == b[0] ? a[1] > b[1] : a[0] < b[0];
        });

        return LIS(envelopes);
    }
};